gnomeExtensions: Some more cleanup

piegamesde opened the pull request:
###### Motivation for this change

- Packages have inconsistent names, let's fix them all https://github.com/NixOS/nixpkgs/pull/122921#issuecomment-846298124
- Add aliases for backwards compatibility with regards to #124295
- Also removed `remove-dropdown-arrows`, because it's old and obsolete.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Added a release notes entry if the change is major or breaking
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

jtojnar commented:
Oh, sorry about that, I did not realize the attribute names are derived from the Nix package names. We will need something like:

```nix
--- a/pkgs/desktops/gnome/extensions/buildGnomeExtension.nix
+++ b/pkgs/desktops/gnome/extensions/buildGnomeExtension.nix
@@ -47,8 +47,11 @@ let
       license = lib.licenses.gpl2Plus; # https://wiki.gnome.org/Projects/GnomeShell/Extensions/Review#Licensing
       maintainers = with lib.maintainers; [ piegames ];
     };
-    # Store the extension's UUID, because we might need it at some places
-    passthru.extensionUuid = uuid;
+    passthru = {
+      extensionPortalSlug = pname;
+      # Store the extension's UUID, because we might need it at some places
+      extensionUuid = uuid;
+    };
   };
 in
   lib.makeOverridable buildGnomeExtension
diff --git a/pkgs/desktops/gnome/extensions/default.nix b/pkgs/desktops/gnome/extensions/default.nix
index f7439897702..1ca9a57eb99 100644
--- a/pkgs/desktops/gnome/extensions/default.nix
+++ b/pkgs/desktops/gnome/extensions/default.nix
@@ -48,7 +48,7 @@ let
     ))
     # Map all extensions to their pname, with potential overwrites
     (map (extension:
-      lib.nameValuePair (extensionRenames.${extension.extensionUuid} or extension.pname) extension
+      lib.nameValuePair (extensionRenames.${extension.extensionUuid} or extension.extensionPortalSlug) extension
     ))
     builtins.listToAttrs
   ];
```

Also, fixup commits for stuff that is in master will cause problems when someone will be rebasing with autosquash.

jtojnar commented on pkgs/desktops/gnome/extensions/draw-on-you-screen/default.nix:
```diff
@@ -1,7 +1,7 @@
 
 stdenv.mkDerivation rec {
-  pname = "gnome-shell-extension-draw-on-your-screen";
+  pname = "gnome-shell-extension-draw-on-you-screen";
```
This looks like a typo.

piegamesde replied:
Yes it very probably is. Nevertheless, this is the "official" name, and thus the name of the automatically packaged extension ¯\_(ツ)_/¯

jtojnar replied:
I would not consider that an official name – it is not used anywhere other than the slug on the extension portal website AFAICT.

piegamesde commented:
Regarding passthru, I just recently realized that the manually packaged extensions simply expose `uuid` in the attrset and people may depend on this. I don't care how we do it out of these two options, but I'd like it to be consistent in some way.

jtojnar commented on pkgs/desktops/gnome/extensions/no-title-bar-forked/default.nix:
```diff
@@ -1,7 +1,7 @@
 
 stdenv.mkDerivation rec {
-  pname = "gnome-shell-extension-no-title-bar";
+  pname = "gnome-shell-extension-no-title-bar-forked";
```
This also sounds like something we should not need.

piegamesde replied:
So my point is, the manually packaged extensions must somehow match the automatically packaged ones in order for overriding to work.

jtojnar replied:
Could we do it in the other direction – rename the automatically packaged extension instead? Since the names are arbitrary, we might as well use ones that make sense.

piegamesde replied:
Middle ground: I'd like to keep the `pname` the same, but I'd be fine with renaming the attribute name of the extensions (pname is almost worthless anyways).

The most promising approach at implementing this would be to key all manually packaged extensions with their UUID as well, so that I can first do the merge/override, and then give them proper names.

jtojnar replied:
`pname` is what Repology uses so we should keep it somewhat canonical and reasonable.

I would say the easiest would be adding a section for changing bad slugs in `extensionRenames.nix`. But adding detection of duplicates based on UUID would be useful when people add a new manually packaged extension so that they choose a correct name.

jtojnar commented:
Pushed the fix for the accidental rename as 571d540abf4.

piegamesde committed:
gnomeExtensions: Remove remove-dropdown-arrows

piegamesde committed:
gnomeExtensions: integrate manually packaged extensions better with the others

Their keys are now not used directly. Instead, we go the standard route of mapping to the
UUID in order to apply the rename procedure. This makes sure the manual override always does
the correct thing, and also gives us more consistency overall.

piegamesde commented:
Second try, I almost started from scratch. No renames just yet, as they are less important with the improved merging. Personally, I'd like to rename all folders of the manually packaged extensions to use their UUID instead of their pname, but let me know what you think about this first.

piegamesde committed:
gnomeExtensions.workspace-matrix: Remove manual packaging

piegamesde committed:
gnomeExtensions.window-is-ready-remover: Remove manual packaging

piegamesde committed:
gnomeExtensions.timepp: Remove manual packaging

piegamesde committed:
gnomeExtensions.noannoyance: Remove manual packaging

piegamesde committed:
gnomeExtensions.mpris-indicator-button: Remove manual packaging

piegamesde committed:
gnomeExtensions.material-shell: Remove manual packaging

piegamesde committed:
gnomeExtensions.hot-edge: Remove manual packaging

piegamesde committed:
gnomeExtensions: Remove gtile

It was never actually referenced anywhere, so technically it was never packaged …

piegamesde committed:
gnomeExtensions.draw-on-your-screen: Remove manual packaging

piegamesde committed:
gnomeExtensions.disable-unredirect: Remove manual packaging

piegamesde committed:
gnomeExtensions.clipboard-indicator: Remove manual packaging

piegamesde commented:
So I went ahead and removed all manual packages that looked like the usual generic boilerplate. @chkno @jtojnar @emanueleperuffo @svsdep @hedning @tu-maurice @benley @Mdsp9070 @ericdallo @alesguzik @eduardosm @jonafato your extensions are affected. Please have a look and comment if any of them can not simply be replaced by the automatic packaging.

piegamesde committed:
gnomeExtensions: rename uuid field to extensionUuid

rhoriguchi commented:
> So I went ahead and removed all manual packages that looked like the usual generic boilerplate. @chkno @jtojnar @emanueleperuffo @svsdep @hedning @tu-maurice @benley @Mdsp9070 @ericdallo @alesguzik @eduardosm @jonafato your extensions are affected. Please have a look and comment if any of them can not simply be replaced by the automatic packaging.

You can also remove `clock-override`, `dynamic-panel-transparency` and  `unite`.

piegamesde commented:
Thanks for the input.

> Edit: I'm not 100% sure about unite because of xprop and clock-override because of gettext. Maybe add some easy way to add some extra buildInputs.

If those are only used at build time, they do not matter. Those building ("building") the extensions on extensions.gnome.org will have done that step for us. The important question is whether or not these dependencies are required at runtime.

jtojnar commented on pkgs/desktops/gnome/extensions/extensionRenames.nix:
```diff
@@ -26,4 +26,42 @@
   #"flypie@schneegans.github.com" = null;
+
+  # These are overrides for the manually packaged ones.
+  "appindicatorsupport@rgcjonas.gmail.com" = "appindicator";
```
Maybe we could even use `passthru.extensionSlug` to make this even more consistent.

piegamesde replied:
You mean for the value of the attributes? This would result in some extensions being renamed (at least their attribute in `nixpkgs`), requiring an alias. As I learned, renaming extensions is a rather tricky topic (since the `extensionSlug` may contain errors), so I thought I'd leave them as they are now and do that some time later (the pnames need fixing as well, for Repology consistency).

Btw these lines cause some of the automatically packaged extensions to be renamed too. I consider this a good thing, because it previously was weird having the automatically and manually packaged extensions available under different attribute names.

jtojnar replied:
I meant that we would add `passthru.extensionSlug` containing manually selected `pname` to the manually packaged extensions and then `mapReadableNames` would use that without the need to define aliases here. But maybe having the names for manually packaged extensions is more sensible approach. Not sure.

piegamesde replied:
Please reiterate. What is exactly your problem here and what do suggest?

jtojnar replied:
The “issue” is that we essentially have two different ways of setting readable names. For automatically packaged extension, it comes from the `passthru.extensionSlug` field, while for the manually packaged, we set the name here. I would suggest unifying it.

But I would not block on this. We can resolve it later if we decide it is a reasonable change.

piegamesde replied:
Oh, now I understand your concern. I pushed a commit that kind of does what you suggested: let me know what you think of it (and good luck trying to catch any potential mishaps/typos). I'll try to clean up the history afterwards.

jtojnar commented on pkgs/desktops/gnome/extensions/EasyScreenCast/default.nix:
```diff
@@ -26,6 +26,7 @@ stdenv.mkDerivation rec {
   makeFlags = [ "INSTALLBASE=$(out)/share/gnome-shell/extensions" ];
 
   uuid = "EasyScreenCast@iacopodeenosee.gmail.com";
```
It should probably be okay to rename the attribute without keeping the old one. This is not really an API we promised to keep; Hyrum, or not.

piegamesde commented on pkgs/desktops/gnome/extensions/default.nix:
```diff
@@ -60,14 +60,17 @@ in rec {
-    (mapReadableNames (produceExtensionsList "40"))
-    // (callPackage ./manuallyPackaged.nix {})
+    (mapReadableNames
+      (lib.attrValues (gnome40Extensions // (import ./manuallyPackaged.nix {inherit callPackage;})))
```
Because I forgot to push that part

eduardosm commented on pkgs/desktops/gnome/extensions/default.nix:
```diff
@@ -71,6 +71,7 @@ in rec {
       nohotcorner = throw "gnomeExtensions.nohotcorner removed since 2019-10-09: Since 3.34, it is a part of GNOME Shell configurable through GNOME Tweaks.";
       mediaplayer = throw "gnomeExtensions.mediaplayer deprecated since 2019-09-23: retired upstream https://github.com/JasonLG1979/gnome-shell-extensions-mediaplayer/blob/master/README.md";
       remove-dropdown-arrows = throw "gnomeExtensions.remove-dropdown-arrows removed since 2021-05-25: The extensions has not seen an update sine GNOME 3.34. Furthermore, the functionality it provides is obsolete as of GNOME 40.";
+      topicons-plus = throw "gnomeExtensions.topicons-plus removed since 2021-25-29: Upstream has abandoned it";
```
There is a PR to enable GNOME 40 support (https://github.com/NixOS/nixpkgs/pull/123392)

piegamesde replied:
Upstream says they don't maintain the project actively anymore. There won't ever be wayland support and they suggest switching to https://github.com/ubuntu/gnome-shell-extension-appindicator instead. But if you are willing to maintain it nevertheless with the patches on top, I can of course keep it in.

eduardosm replied:
I will continue maintaining it. Unfortunately, there are still applications that still use the legacy X11 tray and won't work with appindicator.

chkno commented:
Automatic packaging looks fine for workspace-matrix.  Thanks for doing this cleanup!

piegamesde committed:
gnomeExtensions.dynamic-panel-transparency: Remove manual packaging

piegamesde committed:
gnomeExtensions.appindicator: Remove manual packaging

wucke13 commented:
Will this be backported to 21.05? Currently we ship gnome4 and many many extensions are not compatible, rendering them essentially useless.

jtojnar reviewed (commented):
Sorry it took me so long to get back to you on this. This is clearly an improvement so let’s go forward with this.

The only issue I see is the `easyScreenCast` renaming.

Also the appindicator thing, how did we decide about that?

piegamesde commented on pkgs/desktops/gnome/extensions/extensionRenames.nix:
```diff
@@ -26,4 +26,42 @@
+  "drawOnYourScreen@abakkk.framagit.org" = "draw-on-your-screen";
+  "drop-down-terminal@gs-extensions.zzrough.org" = "drop-down-terminal";
+  "dynamic-panel-transparency@rockon999.github.io" = "dynamic-panel-transparency";
+  "EasyScreenCast@iacopodeenosee.gmail.com" = "EasyScreenCast";
```
It's been a while, but I cannot find any reason for this name anymore. So maybe I meant it to be `easyScreenCast` in the first place and this was just a typo?

piegamesde replied:
Fixed in latest commit.

piegamesde commented:
@wucke13 This PR does not do any actual updates, see [can't find PR right now, sorry] for the actual updates. As it only does refactoring, I'm kind of reluctant to backport it in the fear of breaking things (already happened not long ago). But on the other side it will probably be required or otherwise all extension updates on stable will have to be done manually & separately from unstable.

jtojnar commented on pkgs/desktops/gnome/extensions/extensionRenames.nix:
```diff
@@ -26,4 +26,16 @@
+  "unredirect@vaina.lt" = "disable-unredirect";
+  "drawOnYourScreen@abakkk.framagit.org" = "draw-on-your-screen";
+  "EasyScreenCast@iacopodeenosee.gmail.com" = "easyScreenCast";
+  "gnome-fuzzy-app-search@gnome-shell-extensions.Czarlie.gitlab.com" = "fuzzy-app-search";
```
Why not do this in passthru as well?

piegamesde replied:
Because I only put the `extensionSlug` in passthru, and these differ from it.

jtojnar replied:
I guess we could add a comment like `extensionPortal slug differs from the attribute name so it is not used.` to the `passthru`. And have these in a separate section `These are manually packaged extensions that have different attribute name from the slug on the extension portal`.

piegamesde replied:
I just noticed/remembered: Most of these extensions are not manually packaged anymore (all except three or something), thus we can't easily override that actually.

jtojnar replied:
Yeah, hence the suggestion to add another section to `extensionRenames`. Or do you mean that the `pname`s will not be changed and thus Repology will be confused?

piegamesde replied:
I'm sorry, but I don't think I understand. Isn't this what we're already doing?

jtojnar replied:
Then I am not sure what you mean by “we can't easily override that actually” :smile_cat:

piegamesde replied:
Okay. Let's have a look at `unredirect@vaina.lt` for example. It is automatically packaged, so its `extensionPortalSlug`, `disable-unredirect-fullscreen-windows` would become its `pname`. The line `"unredirect@vaina.lt" = "disable-unredirect";` changes that to `disable-unredirect`. That name is (arguably) more sensible than the other one, but also it is the one we previously had and we need to keep it by for backwards compatibility.

jtojnar replied:
That only affects the attribute path, not the derivation name:

```
nix-repl> gnomeExtensions.disable-unredirect
«derivation /nix/store/xwhialxgr3qpfwvn4r0mbzmzzdlcsyr1-gnome-shell-extension-disable-unredirect-fullscreen-windows-3.drv»
```

piegamesde replied:
Yes, because of how the pipeline works. Is this a problem?

jtojnar replied:
It might be an issue for Repology, depending on how other repositories name the extension. But we can fix that separately since this was already an issue before this PR.

jtojnar commented on pkgs/desktops/gnome/extensions/extensionRenames.nix:
```diff
@@ -26,4 +26,16 @@
+  "gnome-fuzzy-app-search@gnome-shell-extensions.Czarlie.gitlab.com" = "fuzzy-app-search";
+  "paperwm@hedning:matrix.org" = "paperwm";
+  "timepp@zagortenay333" = "timepp";
+  "TopIcons@phocean.net" = "topicons-plus";
```
And this one.

piegamesde replied:
Makes sense, added a comment to all of them.

jtojnar commented on pkgs/desktops/gnome/extensions/extensionRenames.nix:
```diff
@@ -26,4 +26,16 @@
   # That extension is broken because of https://github.com/NixOS/nixpkgs/issues/118612
   #"flypie@schneegans.github.com" = null;
+
+  # These are overrides for the manually packaged ones (some of which got deleted – as in: replaced with
+  # some automatic packaging –, but we still need the "old" name for backwards compatibility).
```
Maybe something like: “These are overrides for extensions that were manually packaged in the past but were replaced by automatic packaging. We preserve the old attribute name for backwards compatibility.”

jtojnar replied:
Though maybe we could just create aliases and switch to the new names (where they do not contain typos).

piegamesde replied:
I just added the extension slug as a comment to all of them. Most of the renames are really sensible and I'd like to keep them. A few could indeed be replaced by an alias, you decide if it should be done in this PR (or if it is even worth the effort).

jtojnar replied:
The switch to aliases does not need to be done in this PR. Just something to keep in mind if we decide that some of the automatic names are more sensible than the old manual ones.

jtojnar commented on pkgs/desktops/gnome/extensions/extensionRenames.nix:
```diff
@@ -26,4 +26,16 @@
+  "appindicatorsupport@rgcjonas.gmail.com" = "appindicator";
+  "unredirect@vaina.lt" = "disable-unredirect";
+  "drawOnYourScreen@abakkk.framagit.org" = "draw-on-your-screen";
+  "EasyScreenCast@iacopodeenosee.gmail.com" = "easyScreenCast";
```
Also here.

jtojnar commented on pkgs/desktops/gnome/extensions/no-title-bar/default.nix:
```diff
@@ -23,7 +23,10 @@ stdenv.mkDerivation rec {
-  uuid = "no-title-bar@jonaspoehler.de";
+  passthru = {
+    extensionUuid = "no-title-bar@jonaspoehler.de";
+    extensionPortalSlug = "no-title-bar";
```
Probably should comment that actual slug is `no-title-bar-forked`

piegamesde replied:
Thank you, good catch! No, I think I should set it to `no-title-bar-forked` and then add it to the renames list.

jtojnar replied:
Should we remove this here then?

piegamesde replied:
Not sure. Actually, I planned to add this to all extensions so that it's more consistent, but I didn't get to / was to lazy and there was no urgent need.

jtojnar replied:
If we add this to all extensions, it either should match the actual slug (mostly useless IMO and could confuse people expecting it to match the attribute name) or it should be the corrected version but then the change should be noted in a comment.

piegamesde replied:
But should we? (Also, I'm starting to think of simply renaming `extensionPortalSlug` to something like `attributeName` to make sure it should only be used for attribute name calculation.)

jtojnar replied:
I do not have a preference either way. The reason I went with `extensionPortalSlug` is precisely because it does not have to be attribute name in the case of automatically packaged extensions – it can be renamed.

piegamesde replied:
I see. Then I suggest keeping it as is for the sake of getting this merged eventually.

jtojnar replied:
The comments or removal would be nice for the cases where there is the discrepancy, like here, though.

jtojnar commented on pkgs/desktops/gnome/extensions/unite/default.nix:
```diff
@@ -11,22 +11,25 @@ stdenv.mkDerivation rec {
-  uuid = "unite@hardpixel.eu";
+  passthru = {
+    extensionUuid = "unite@hardpixel.eu";
+    extensionPortalSlug = "unit";
```
This should be `unite`.

jtojnar commented on pkgs/desktops/gnome/extensions/extensionRenames.nix:
```diff
@@ -26,4 +26,17 @@
+  # These are overrides for extensions that were manually packaged in the past but are gradually
+  # being replaced by automatic packaging where possible. We preserve the old attribute name for
+  # backwards compatibility.
+  "appindicatorsupport@rgcjonas.gmail.com" = "appindicator"; # upstream is "appindicator-support"
```
I would replace upstream with portal slug since upstream is ambiguous here.

piegamesde commented:
Let's summarize the discussion so far because the next time I'll have forgotten most of it:

- We should take a look a the `pname`s and compare with Repology
- The `extensionPortalSlug` can be normalized some more.
- (list will be expanded)

jtojnar commented on pkgs/desktops/gnome/extensions/extensionRenames.nix:
```diff
@@ -26,4 +28,24 @@
+  "gnome-fuzzy-app-search@gnome-shell-extensions.Czarlie.gitlab.com" = "fuzzy-app-search"; # extensionPortalSlug is "gnome-fuzzy-app-search"
+  "TopIcons@phocean.net" = "topicons-plus"; # extensionPortalSlug is "topicons"
+  "paperwm@hedning:matrix.org" = "paperwm"; # is not on extensions.gnome.org
+  "no-title-bar@jonaspoehler.de" = "no-title-bar" # extensionPortalSlug is "no-title-bar-forked"
```
Missing semicolon

piegamesde replied:
I'm sorry, it's late.

jtojnar commented on pkgs/desktops/gnome/extensions/no-title-bar/default.nix:
```diff
@@ -23,7 +23,10 @@ stdenv.mkDerivation rec {
-  uuid = "no-title-bar@jonaspoehler.de";
+  passthru = {
+    extensionUuid = "no-title-bar@jonaspoehler.de";
+    extensionPortalSlug = "no-title-bar-forked";
```
For consistency with other manual extensions that do not match, I would just drop this line.

jtojnar reviewed (approved):
LGTM. Thanks for bearing with me.

piegamesde commented:
Thanks for all the reviews
